Lookup-table-based light simulation

peter-madigan opened the pull request:
Same as PR #69 except with a few revisions based on review comments.

Also cleaned up merge conflicts.

soleti commented on cli/dumpTree.py:
```diff
@@ -175,8 +210,8 @@ def dump(input_file, output_file):
             trajectories[iTraj]["xyz_end"] = (end_pt.GetPosition().X(), end_pt.GetPosition().Y(), end_pt.GetPosition().Z())
-            trajectories[iTraj]["t_start"] = start_pt.GetPosition().T()
-            trajectories[iTraj]["t_end"] = end_pt.GetPosition().T()
+            trajectories[iTraj]["t_start"] = start_pt.GetPosition().T() / 1e3
```
Could we clarify the unit here just for reference?

soleti commented on cli/dumpTree.py:
```diff
@@ -140,18 +142,51 @@ def dump(input_file, output_file):
 
-    for jentry in range(entries):
-        print(jentry)
+    for jentry in tqdm(range(entries+1)):
```
Why the +1 here?

peter-madigan replied:
This is because I made it so it saves a batch of trajectories every now and then within the loop. For the last iteration, it will always save the batch to the file. I could make this less weird if you'd like (just make a dedicated function for writing to the file and then call it after the loop exits).

soleti commented on cli/simulate_pixels.py:
```diff
@@ -221,30 +242,42 @@ def run_simulation(input_filename,
     step = 1
+    
+    # charge simulation
+    event_id_list = []
```
Can you clarify the logic here? I had moved these inside the for in order to save the output after each event, does it now happen every `EVENT_BATCH_SIZE` events?

peter-madigan replied:
That's right. It helps speed up the Module 0 simulation a bit, since each event isn't actually that big. That can be adjusted to optimize for different detectors (or made into a script parameter). 

As a side note, I did a little bit of profiling and it now spends about equal time in the "simulation" piece as in the "export_to_hdf5" piece. So, there's a potential x2 speed-up that could be achieved by letting the export to hdf5 run in a separate process. Just something to think about - it would take a fair bit of work.

soleti commented on cli/simulate_pixels.py:
```diff
@@ -333,7 +366,7 @@ def run_simulation(input_filename,
             RangePush("sum_pixels_signals")
             # Here we combine the induced current on the same pixels by different tracks
-            TPB = (8,8,8)
+            TPB = (1,1,64)
```
This is because it's faster?

peter-madigan replied:
I guess - this was not really intended in this PR. I think you maybe mentioned it a while ago, and I tried running with these parameters. I didn't notice a difference, but yeah this was an accident.

soleti commented on larndsim/consts/light.py:
```diff
@@ -34,5 +83,33 @@ def set_light_properties(detprop_file):
+                if idet in tpc_to_op_channel[itpc]:
+                    OP_CHANNEL_TO_TPC[idet] = itpc
+
+        LIGHT_TICK_SIZE = float(detprop.get('light_tick_size', LIGHT_TICK_SIZE))
```
I see these are now loaded from the YAML, if present, otherwise they assume the default values. However, it would be nice for reference to have these values in the module0 YAML file

soleti commented on larndsim/drifting.py:
```diff
@@ -29,7 +29,7 @@ def drift(tracks):
         track = tracks[itrk]
 
-        pixel_plane = -1
+        pixel_plane = detector.DEFAULT_PLANE_INDEX
```
Thank you for this, much better than -1

peter-madigan replied:
Yes thanks Dan!

soleti reviewed (commented):
Looks good to me, apart from some minor comments. Am I right to assume there is no MC truth - light simulation association atm?

peter-madigan commented:
OK @soleti take a look now. I think I've addressed all your comments.

The only truth information that is saved is the intensity data for each track segment, which does require some interpretation. But it's not obvious to me what other information we would want that isn't just the sample-by-sample truth info (which quickly becomes a ton of mostly irrelevant data). I figure we can start with what we have, and if later on we find out there is additional information that would be good to track, deal with it then.
